src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,107 @@
 package core.basesyntax;
 
+import core.basesyntax.util.ObjectUtil;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static int capacity = 10;
+    private static final double RESIZE_FACTOR = 1.5;
+    private int size;
+    private T[] elements = (T[]) new Object[capacity];
+
     @Override
     public void add(T value) {
-
+        checkCapacity();
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkAddAtIndex(index);
+        checkCapacity();
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        final T removedElement = elements[index];
+        System.arraycopy(elements, index + 1, elements, index, size - index - 1);
+        elements[size - 1] = null;
+        size--;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = findElement(element);
+        return remove(index);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkCapacity() {
+        if (capacity == size) {
+            increaseCapacity();
+        }
+    }
+
+    private void increaseCapacity() {
+        capacity *= RESIZE_FACTOR;
```

Comment:
create here temporary variable.
```suggestion
        int capacity = elements.length * RESIZE_FACTOR;
```